lib/ipam/ipam_test.go

```diff
@@ -600,6 +600,128 @@ var _ = testutils.E2eDatastoreDescribe("IPAM tests", testutils.DatastoreAll, fun
 		})
 	})
 
+	Describe("IPAM Pool Details", func() {
+
+		AfterEach(func() {
+			deleteAllPools()
+		})
+
+		It("should get an IP from pool1 when explicitly requesting from that pool", func() {
+			bc.Clean()
+			deleteAllPools()
+
+			pool1 := cnet.MustParseNetwork("10.0.0.0/24")
+			host := "host-a"
+
+			err := applyNode(bc, kc, host, nil)
+			Expect(err).NotTo(HaveOccurred())
+			applyPool("10.0.0.0/24", true, "")
+			applyPool("20.0.0.0/24", true, "")
+
+			args := AutoAssignArgs{
+				Num4:      1,
+				Num6:      0,
+				Hostname:  host,
+				IPv4Pools: []cnet.IPNet{pool1},
+			}
+
+			v4, _, outErr := ic.AutoAssign(context.Background(), args)
+			Expect(outErr).NotTo(HaveOccurred())
+			Expect(pool1.IPNet.Contains(v4[0].IP)).To(BeTrue())
+		})
+	})
+
+	Describe("IPAM IP stealing", func() {
+		node1 := "host1"
+		node2 := "host2"
+
+		AfterEach(func() {
+			ctx := context.Background()
+			cfg := IPAMConfig{AutoAllocateBlocks: true, StrictAffinity: false}
+			err := ic.SetIPAMConfig(ctx, cfg)
+			Expect(err).NotTo(HaveOccurred())
+
+			deleteAllPools()
+		})
+
+		It("should get IP when ip stealing is enabled", func() {
+			bc.Clean()
+			deleteAllPools()
+
+			err := applyNode(bc, kc, node1, map[string]string{"foo": "bar"})
+			Expect(err).NotTo(HaveOccurred())
+
+			err = applyNode(bc, kc, node2, map[string]string{"foo": "bar"})
+			Expect(err).NotTo(HaveOccurred())
+
+			// Only one block can be created out of the pool.
+			// When ip stealing is enabled, both nodes will be able to assign IP addresses
+			// When ip stealing is disabled only one node, the ones that gets the block, will
+			// be able to assign IP address
+			applyPoolWithBlockSize("10.0.0.0/29", true, `foo == "bar"`, 29)
+
+			v4, _, err := ic.AutoAssign(context.Background(), AutoAssignArgs{
+				Num4:     1,
+				Num6:     0,
+				Hostname: node1,
+			})
+			Expect(err).ToNot(HaveOccurred())
+			Expect(len(v4)).To(Equal(1))
+
+			v4, _, err = ic.AutoAssign(context.Background(), AutoAssignArgs{
+				Num4:     1,
+				Num6:     0,
+				Hostname: node2,
+			})
+			Expect(err).ToNot(HaveOccurred())
+			Expect(len(v4)).To(Equal(1))
+		})
+
+		It("should get no IP when ip stealing is disabled and no IP are available for node", func() {
```

Comment:
"no IP are available for node" -> "no blocks are available for the node"

Reply:
It might be a good idea to combine these two tests - i.e., show that you can modify the IPAMConfig and that the changes take effect.

e.g.,

- Allocate an IP
- Borrow an IP, show it works
- Enable StrictAffinity
- Borrow an IP, show it fails
- Disable StrictAffinity
- Borrow an IP, show it works